internal/io.h

```diff
@@ -8,8 +8,123 @@
  *             file COPYING are met.  Consult the file for details.
  * @brief      Internal header for IO.
  */
+
+#define RB_IO_T
+
 #include "ruby/ruby.h"          /* for VALUE */
-#include "ruby/io.h"            /* for rb_io_t */
+#include "ruby/io.h"
+
+/**
+ * IO  buffers.   This  is  an implementation  detail  of  ::rb_io_t::wbuf  and
+ * ::rb_io_t::rbuf.  People don't manipulate it directly.
+ */
+PACKED_STRUCT_UNALIGNED(struct rb_io_internal_buffer {
+    /** Pointer to the underlying memory region, of at least `capa` bytes. */
+    char *ptr;                  /* off + len <= capa */
+
+    /** Offset inside of `ptr`. */
+    int off;
+
+    /** Length of the buffer. */
+    int len;
+
+    /** Designed capacity of the buffer. */
+    int capa;
+});
+
+/** Ruby's IO, metadata and buffers. */
+struct rb_io {
+    /** The IO's Ruby level counterpart. */
+    VALUE self;
+
+    /** stdio ptr for read/write, if available. */
+    FILE *stdio_file;
+
+    /** file descriptor. */
+    int fd;
+
+    /** mode flags: FMODE_XXXs */
+    int mode;
+
+    /** child's pid (for pipes) */
+    rb_pid_t pid;
+
+    /** number of lines read */
+    int lineno;
+
+    /** pathname for file */
+    VALUE pathv;
+
+    /**
+     * Duplex IO object, if set.
+     *
+     * @see rb_io_set_write_io()
+     */
+    VALUE tied_io_for_writing;
+
```

Comment:
Probably needs a comment here and in ruby/io.h to keep the fields above in sync.

Reply:
Done, good call out.